### PR TITLE
Milestone 2 of ML pipeline migration

### PR DESCRIPTION
**Pull request by prasanna08**
This PR introduces changes related to [milestone 2](https://docs.google.com/document/d/1fIlDem4NC6eqEHwjkn2FNdHuo6eEsg3LMLd2QAfxCPo/edit#heading=h.ftzctx55cmhv) of ML pipeline migration.

This milestone consists of PRs on both [Oppia](https://github.com/oppia/oppia/pull/10060) as well as OppiaML since it affects the communication paradigm between the two. Hence, they should be merged simultaneously, so that those changes requested in one PR affecting the other can be resolved together at the same time. Simultaneous merging isn't absolute necessity but it avoids opening another PR to address any further changes which may be required if one is merged before the other (for e.g. Suppose that OppiaML PR is merged earlier but later on some significant change on Oppia PR may require to fix things in OppiaML as well).

### CODE REVIEW

File: .pylintrc [hunk 1]
```diff
@@ -66,7 +66,7 @@ ignore-imports=yes
 #    too-many-statements
 # and fix those issues.
 
-disable=duplicate-code
+disable=duplicate-code, no-member
```
Comment (seanlip) [inline .pylintrc:72]: Why add this? Try not to.

File: core/classifiers/CodeClassifier/CodeClassifier.py [hunk 2]
```diff
@@ -538,6 +538,10 @@ def __init__(self):
         # text into a feature a vector.
         self.count_vector = None
 
+    @property
+    def version(self):
+        return 1
```
Comment (seanlip) [inline core/classifiers/CodeClassifier/CodeClassifier.py]: Should 1 be a global constant?

File: core/classifiers/CodeClassifier/CodeClassifier.py [hunk 3]
```diff
@@ -709,7 +716,7 @@ def train(self, training_data):
 
     # pylint: enable=too-many-locals
     # pylint: disable=too-many-branches, no-self-use
-    def validate(self, classifier_data):
+    def validate(self, classifier_data): #pylint: disable=arguments-differ
```
Comment (seanlip) [inline core/classifiers/CodeClassifier/CodeClassifier.py]: Try to avoid the pylint pragma?

File: core/classifiers/TextClassifier/TextClassifier.py [hunk 4]
```diff
@@ -116,29 +121,36 @@ def train(self, training_data):
...
     # pylint: disable=too-many-branches
-    def validate(self, classifier_data):
-        """Validates classifier data.
+    def validate(self, frozen_model_proto):
+        """Validates frozen model of trained classifier.
 
         Args:
-            classifier_data: dict of the classifier attributes specific to
-                the classifier algorithm used.
+            frozen_model_proto: TextClassifierFrozenModel object consisting of
```
Comment (seanlip) [inline core/classifiers/TextClassifier/TextClassifier.py]: Follow usual args docstring practices: `arg_name: arg_type. Arg description.`

File: core/domain/job_services.py [hunk 5]
```diff
@@ -118,15 +135,9 @@ def store_job_result(job_id, classifier_data):
     # Therefore, converting all floating point numbers to string keeps
```
Comment (seanlip) [inline core/domain/job_services.py]: Is the comment still relevant?

File: core/domain/remote_access_services.py [hunk 6]
```diff
@@ -62,16 +64,17 @@ def _get_shared_secret():
...
 
 
-def generate_signature(data):
-    """Generates digital signature for given data.
+def generate_signature(message, vm_id):
+    """Generates digital signature for given message combined with vm_id.
 
     Args:
-        data: dict. A dictionary data to be transferred over network.
+        messate: str. Message string.
```
Comment (seanlip) [inline core/domain/remote_access_services.py]: Typo in argname

**Review by seanlip — CHANGES_REQUESTED** [review]
Thanks! Took first pass.

File: core/classifiers/base.py [hunk 7]
```diff
@@ -36,6 +36,14 @@ class BaseClassifier(object):
     def __init__(self):
         pass
 
+    @property
+    @abc.abstractproperty
+    def version(self):
+        """Version of the classifier algorithm. Version of algorithm is matched
```
Comment (BenHenning) [inline core/classifiers/base.py]: Nit on wording: ``The version  of the algorithm is matched...``

File: core/classifiers/base.py [hunk 8]
```diff
@@ -84,11 +93,12 @@ def train(self, training_data):
         raise NotImplementedError
 
     @abc.abstractmethod
-    def validate(self, classifier_data):
-        """Validates classifier data.
+    def validate(self, frozen_model_proto):
+        """Validates frozen model protobuf object containing parameters of
```
Comment (BenHenning) [inline core/classifiers/base.py]: Nit on wording: ``Validates the specified frozen model protobuf...``

File: core/domain/job_services.py [hunk 9]
```diff
@@ -96,19 +107,25 @@ def train_classifier(algorithm_id, training_data):
     """
     classifier = algorithm_registry.Registry.get_classifier_by_algorithm_id(
         algorithm_id)
+    if classifier.version != algorithm_version:
+        logging.warning(
+            'Classifier version %d mismatches algorithm version %d received '
+            'in job data', classifier.version, algorithm_version)
+        return None
```
Comment (BenHenning) [inline core/domain/job_services.py]: Is there a test to verify that this condition is properly hit as expected?

File: core/domain/training_job_result_domain_test.py [hunk 10]
```diff
@@ -41,6 +41,24 @@ def test_validate_job_data_with_invalid_model_raises_exception(self):
         """Ensure that validation checks raise exception when
         an invalid classifier model is supplied.
         """
+        job_id = 123
+        algorithm_id = 'TextClassifier'
+        classifier_data = 'simple classifier'
+        job_result = training_job_result_domain.TrainingJobResult(
+            job_id, algorithm_id, classifier_data)
+        with self.assertRaisesRegexp(
```
Comment (BenHenning) [inline core/domain/training_job_result_domain_test.py:49]: I suggest splitting up this test to have a targeted validate test for each behavior being checked, rather than clumping them together. This helps better isolate failures when the tests do fail, makes maintaining tests a bit easier, and allows surveying the list of test names to audit the behavioral coverage of the test suite.

**Review by BenHenning — CHANGES_REQUESTED** [review]
Thanks @prasanna08! Just had a few comments beyond the ones Sean left.

File: .pylintrc (hunk 1 above)
Comment (prasanna08) [reply .pylintrc:72]: The problem is with [protobuf generated class objects](https://github.com/PyCQA/pylint/issues/1864). These classes are generated in a way that no explicit member is defined on them (you can use anything for attribute name). Thus anytime we access a pylint object's attributes, it yields the error of `no-member`. Pylint devs have denied adding any feature for this and so either we disable it at every line in the code wherever we access it or globally. Since this is disabled globally in Oppia, I decided to do so here, too.

File: core/classifiers/CodeClassifier/CodeClassifier.py (hunk 2 above)
Comment (prasanna08) [reply core/classifiers/CodeClassifier/CodeClassifier.py]: I don't think so. This version will change with new iteration of classifiers.

File: core/classifiers/CodeClassifier/CodeClassifier.py (hunk 3 above)
Comment (prasanna08) [reply core/classifiers/CodeClassifier/CodeClassifier.py]: This is necessary because the code-classifier doesn't have a proto definition. Hence, we can't use the one defined in the base classifier and need to override the arguments.

File: core/classifiers/TextClassifier/TextClassifier.py (hunk 4 above)
Comment (prasanna08) [reply core/classifiers/TextClassifier/TextClassifier.py]: Done

File: core/domain/job_services.py (hunk 5 above)
Comment (prasanna08) [reply core/domain/job_services.py]: Removed

File: core/domain/remote_access_services.py (hunk 6 above)
Comment (prasanna08) [reply core/domain/remote_access_services.py]: Fixed

File: core/classifiers/base.py (hunk 7 above)
Comment (prasanna08) [reply core/classifiers/base.py]: Done

File: core/classifiers/base.py (hunk 8 above)
Comment (prasanna08) [reply core/classifiers/base.py]: Done

File: core/domain/job_services.py (hunk 9 above)
Comment (prasanna08) [reply core/domain/job_services.py]: Actually, there was no test file for `job_services` module. I just added one with a bunch of test cases for every function and every possible scenarion

File: core/domain/training_job_result_domain_test.py (hunk 10 above)
Comment (prasanna08) [reply core/domain/training_job_result_domain_test.py:49]: Done

File: core/classifiers/base.py [hunk 11]
```diff
@@ -39,9 +39,10 @@ def __init__(self):
     @property
     @abc.abstractproperty
     def version(self):
-        """Version of the classifier algorithm. Version of algorithm is matched
-        with the version received as part of job data before training the
-        classifier."""
+        """Version of the classifier algorithm. The version of algorithm is
```
Comment (BenHenning) [inline core/classifiers/base.py]: ``The version of the algorithm``

File: core/classifiers/base.py [hunk 12]
```diff
@@ -94,8 +95,8 @@ def train(self, training_data):
 
     @abc.abstractmethod
     def validate(self, frozen_model_proto):
-        """Validates frozen model protobuf object containing parameters of
-        trained classifier model.
+        """Validates the specified frozen model protobuf object containing
+        parameters of trained classifier model.
```
Comment (BenHenning) [inline core/classifiers/base.py]: ``parameters of the trained classifier model.``

File: core/domain/training_job_result_domain_test.py [hunk 13]
```diff
@@ -37,9 +37,9 @@ def test_validate_job_data_with_valid_model_does_not_raise_exception(self): # py
             job_id, algorithm_id, classifier_data)
         job_result.validate()
 
-    def test_validate_job_data_with_invalid_model_raises_exception(self):
-        """Ensure that validation checks raise exception when
-        an invalid classifier model is supplied.
+    def test_job_data_with_invalid_job_id_raises_exception_on_validtion(self):
```
Comment (BenHenning) [inline core/domain/training_job_result_domain_test.py]: Optional, but you can also rearrange the test names as so:

``test_validate_with_invalid_job_id_raises_exception``

This keeps the format _test\_action\_with\_condition\_and\_condition\_has\_outcome_. Note that I think the 'job_data' bit is implied given the class name.

Comment (BenHenning) [inline core/domain/training_job_result_domain_test.py]: Typo in ``validation``. Ditto below.

File: core/domain/training_job_result_domain_test.py [hunk 14]
```diff
@@ -37,9 +37,9 @@ def test_validate_job_data_with_valid_model_does_not_raise_exception(self): # py
...
+        """Ensure that validation checks raise exception when an invalid
```
Comment (BenHenning) [inline core/domain/training_job_result_domain_test.py:41]: Here & below: probably don't need the doc comment anymore since the test name is pretty straightforward.

File: core/domain/job_services_test.py [hunk 15]
```diff
@@ -0,0 +1,332 @@
...
+                'Expected answer_group_index to be an int'):
+                job_services.get_next_job()
+
+        job_data['training_data'] = [{
+            'answer_group_index': 1,
+            'answers': 'answer1'
+        }]
+        post_callback = self._get_post_callback_for_next_job(job_data)
+        with self.set_job_request_post_callback(post_callback):
+            with self.assertRaisesRegexp(
```
Comment (BenHenning) [inline core/domain/job_services_test.py:282]: Suggest splitting up the tests in here, as well. This test seems to be verifying a bunch of different behaviors--these would be a bit cleaner if in separate tests.

File: core/domain/job_services_test.py [hunk 16]
```diff
@@ -0,0 +1,332 @@
...
+            'algorithm_version': 1
+        }
+        post_callback = self._get_post_callback_for_next_job(job_data)
+
+        with self.set_job_request_post_callback(post_callback):
+            with self.assertRaisesRegexp(
+                Exception, 'Invalid algorithm id ab'):
+                job_services.get_next_job()
+
+    def test_job_validation_raises_exception_on_unknown_algorithm_version(self):
```
Comment (BenHenning) [inline core/domain/job_services_test.py]: Maybe instead: 'incorrect' version would be clearer?

Comment (BenHenning) [inline core/domain/job_services_test.py]: Suggest also adding a test where the validation passes without an exception being raised.

**Review by BenHenning — COMMENTED** [review]
Thanks @prasanna08! PR mostly LGTM, just had a few more nits.

File: core/domain/job_services.py (hunk 9 above)
Comment (BenHenning) [reply core/domain/job_services.py]: Thanks!

File: core/domain/training_job_result_domain_test.py (hunk 14 above)
Comment (prasanna08) [reply core/domain/training_job_result_domain_test.py:41]: It becomes necessary because of the pylint. The doc comment option hasn't been disabled globally.

File: core/classifiers/base.py (hunk 11 above)
Comment (prasanna08) [reply core/classifiers/base.py]: Done.

File: core/classifiers/base.py (hunk 12 above)
Comment (prasanna08) [reply core/classifiers/base.py]: Done

File: core/domain/training_job_result_domain_test.py (hunk 13 above)
Comment (prasanna08) [reply core/domain/training_job_result_domain_test.py]: Done

Comment (prasanna08) [reply core/domain/training_job_result_domain_test.py]: Done

File: core/domain/job_services_test.py (hunk 15 above)
Comment (prasanna08) [reply core/domain/job_services_test.py:282]: Done

File: core/domain/job_services_test.py (hunk 16 above)
Comment (prasanna08) [reply core/domain/job_services_test.py]: Done

Comment (prasanna08) [reply core/domain/job_services_test.py]: Done

File: .pylintrc (hunk 1 above)
Comment (seanlip) [reply .pylintrc:72]: Sounds reasonable. Maybe add a comment to explain? Would be useful for future devs who wonder the same thing.

File: core/classifiers/CodeClassifier/CodeClassifier.py (hunk 2 above)
Comment (seanlip) [reply core/classifiers/CodeClassifier/CodeClassifier.py]: Ah yes good point, OK thanks.

That said -- maybe call it schema_version? In the main oppia codebase we distinguish between version (of the instance) and schema version (of the field definitions).

Also why make it a property rather than, say, a class constant? Do we expect it to be changed in the future? Maybe there should be a docstring explaining the expected lifecycle and what developers should do / not do.

File: core/classifiers/CodeClassifier/CodeClassifier.py (hunk 3 above)
Comment (seanlip) [reply core/classifiers/CodeClassifier/CodeClassifier.py]: Then add an explanation in the code for that. Also add single space after `#`.

(In general, when you're doing something whose reason is unclear, safer to just add a comment.)

**Review by seanlip — CHANGES_REQUESTED** [review]

File: core/domain/training_job_result_domain_test.py (hunk 14 above)
Comment (BenHenning) [reply core/domain/training_job_result_domain_test.py:41]: Ah, thanks for clarifying.

File: core/domain/job_services_test.py [hunk 17]
```diff
@@ -69,7 +69,20 @@ def test_next_job_is_fetched_correctly(self):
         self.assertEqual(job_data['algorithm_version'], 1)
         self.assertEqual(job_data['training_data'], [])
 
-    def test_job_validation_raises_exception_on_incorrect_job_data_type(self):
+    def test_validate_works_correctly(self):
```
Comment (BenHenning) [inline core/domain/job_services_test.py]: Nit on naming: aren't all of these tests cumulatively verifying that validate works correctly? Perhaps:

``test_validate_with_valid_job_data_raises_no_exceptions``

**Review by BenHenning — APPROVED** [review]
Thanks @prasanna08! Aside from Sean's outstanding comments, this LGTM.

File: core/classifiers/base.py [hunk 18]
```diff
@@ -53,12 +62,13 @@ def type_in_job_result_proto(self):
...
-    def to_dict(self):
-        """Returns a dict representing this classifier.
+    def to_proto(self):
+        """Returns a protobuf of the frozen model consisting of trained
+        parameters.
 
         Returns:
-            dict. A dictionary representation of classifier referred as
-            'classifier_data'. This data is used for prediction.
+            Object. A protobuf object of frozen model containing trained
```
Comment (seanlip) [inline core/classifiers/base.py]: Just checking -- here and elsewhere, is "Object" the actual type? Figured I'd double-check since the name seems rather generic.

Perhaps worth changing the return type to "something.Object" if that helps disambiguate (again, here and elsewhere), based on where the type is defined.

File: core/domain/remote_access_services.py [hunk 19]
```diff
@@ -62,16 +64,17 @@ def _get_shared_secret():
...
 
-def generate_signature(data):
-    """Generates digital signature for given data.
+def generate_signature(message, vm_id):
+    """Generates digital signature for given message combined with vm_id.
 
     Args:
-        data: dict. A dictionary data to be transferred over network.
+        message: str. Message string.
+        vm_id: string. ID of the VM that trained the job.
```
Comment (seanlip) [inline core/domain/remote_access_services.py]: str.

File: core/domain/remote_access_services_test.py [hunk 20]
```diff
@@ -25,15 +28,12 @@ class RemoteAccessServicesTests(test_utils.GenericTestBase):
 
     def test_that_generate_signature_works_correctly(self):
         """Test that generate signature function is working as expected."""
-        data = {
-            'vm_id': 'vm_default'
-        }
-
         with self.swap(vmconf, 'DEFAULT_VM_SHARED_SECRET', '1a2b3c4e'):
-            sign = remote_access_services.generate_signature(data)
+            sign = remote_access_services.generate_signature(
```
Comment (seanlip) [inline core/domain/remote_access_services_test.py]: Call it "signature", here and elsewhere. "sign" means something different.

In genera, try not to abbreviate, especially where you don't need to.

File: core/domain/remote_access_services_test.py [hunk 21]
```diff
@@ -58,62 +59,38 @@ def post_callback(request):
...
+            self.assertEqual(
+                payload.job_result.text_classifier.model_json, 'model_json')
 
         with self.set_job_result_post_callback(post_callback):
             status = remote_access_services.store_trained_classifier_model(
-                job_result_dict)
+                job_result)
         self.assertEqual(status, 200)
-
-    def test_exception_is_raised_when_classifier_data_is_inappropriate(self):
```
Comment (seanlip) [inline core/domain/remote_access_services_test.py:94]: Just to check, why was this deleted? Isn't it still possible for job_result.validate() to fail somehow?

(I know you have tests for it specifically, but the point is to make sure that validate() is called from this function too in order to ensure that invalid models can't be saved.)

File: core/domain/job_services.py [hunk 22]
```diff
@@ -83,11 +100,12 @@ def get_next_job():
...
 
-def train_classifier(algorithm_id, training_data):
+def train_classifier(algorithm_id, algorithm_version, training_data):
     """Train classifier associated with 'algorithm_id' using 'training_data'.
 
     Args:
         algorithm_id: str. ID of classifier algorithm.
+        algorithm_version: int. Version of the classifier algorithm.
         training_data: list(dict). A list containing training data. Each dict
             stores 'answer_group_index' and 'answers'.
```
Comment (seanlip) [inline core/domain/job_services.py:110]: Is the return value still correct?

File: main.py [hunk 23]
```diff
@@ -36,15 +36,19 @@ def main():
...
-            job_data['job_id'], classifier_data)
+        frozen_model_proto = job_services.train_classifier(
+            job_data['algorithm_id'], job_data['algorithm_version'],
+            job_data['training_data'])
 
-        if status != 200:
-            logging.warning(
-                'Failed to store result of the job with \'%s\' job_id',
-                job_data['job_id'])
+        if frozen_model_proto:
```
Comment (seanlip) [inline main.py]: Under what circumstances would this not hold?

**Review by seanlip — CHANGES_REQUESTED** [review]
(Left some additional comments.)

File: main.py (hunk 23 above)
Comment (prasanna08) [reply main.py]: It will hold for all the cases now. Earlier the `algorithm_version` check happened when training the classifier (and it would hold if `algoritm_version` had a mismatch) but now I moved it to job validation checks. Hence, this condition is removed now as it will always hold.

File: core/domain/job_services.py (hunk 22 above)
Comment (prasanna08) [reply core/domain/job_services.py:110]: Fixed.

File: core/domain/remote_access_services_test.py (hunk 21 above)
Comment (prasanna08) [reply core/domain/remote_access_services_test.py:94]: Added a complimentary test case for this, it does not do full set of validation checks but just ensures that validate has been called.

File: core/domain/remote_access_services_test.py (hunk 20 above)
Comment (prasanna08) [reply core/domain/remote_access_services_test.py]: Done

File: core/domain/remote_access_services.py (hunk 19 above)
Comment (prasanna08) [reply core/domain/remote_access_services.py]: Done

File: core/classifiers/base.py (hunk 18 above)
Comment (prasanna08) [reply core/classifiers/base.py]: Using the `FrozenModel` definition as defined in [#10060](https://github.com/oppia/oppia/pull/10060). Also added README in `core/domain/proto`

File: core/domain/job_services_test.py (hunk 17 above)
Comment (prasanna08) [reply core/domain/job_services_test.py]: Renamed.

**Comment by seanlip** [comment]
@prasanna08 I think you haven't replied to all the comments yet? Please double-check the full conversation history and reassign to me/Ben once everything is fully addressed. Thanks!

File: .pylintrc (hunk 1 above)
Comment (prasanna08) [reply .pylintrc:72]: Added a comment,

File: core/classifiers/CodeClassifier/CodeClassifier.py (hunk 2 above)
Comment (prasanna08) [reply core/classifiers/CodeClassifier/CodeClassifier.py]: We can call it `algorithm_version`, schema is something that evolves over time and is backward compatible, however, classifiers won't be backward compatible and need to be recreated. So we can use `algorithm_version` to refer to this. And yes, this is expected to change in the future with new versions of classifiers.

Comment (prasanna08) [reply core/classifiers/CodeClassifier/CodeClassifier.py]: Changed to `algorithm_version`.

File: core/classifiers/CodeClassifier/CodeClassifier.py (hunk 3 above)
Comment (prasanna08) [reply core/classifiers/CodeClassifier/CodeClassifier.py]: Added a comment.

**Comment by prasanna08** [comment]
Hi @seanlip, I missed 3 older comments, my bad. Fixed now.

File: .pylintrc (hunk 1 above)
Comment (seanlip) [reply .pylintrc:72]: Thanks -- but please add the comment *above* the line in question, not below it.

File: core/classifiers/CodeClassifier/CodeClassifier.py (hunk 2 above)
Comment (seanlip) [reply core/classifiers/CodeClassifier/CodeClassifier.py]: Thanks. Maybe you can add a docstring explaining what it is and that it is expected to change in a non-backwards-compatible way (I know you have this in the parent class but it helps to see this or the relevant subset of it in the subclasses too).

File: core/classifiers/CodeClassifier/CodeClassifier.py (hunk 3 above)
Comment (seanlip) [reply core/classifiers/CodeClassifier/CodeClassifier.py]: Add single space after #.

File: core/classifiers/CodeClassifier/CodeClassifier.py [hunk 24]
```diff
@@ -709,7 +716,12 @@ def train(self, training_data):
 
     # pylint: enable=too-many-locals
     # pylint: disable=too-many-branches, no-self-use
-    def validate(self, classifier_data):
+    # CodeClassifier has been deprecated as it does not support protobuf based
```
Comment (seanlip) [inline core/classifiers/CodeClassifier/CodeClassifier.py]: If it's deprecated, should we just delete the whole thing rather than keep the code around just to support the test cases?

File: core/classifiers/base.py [hunk 25]
```diff
@@ -36,6 +36,20 @@ class BaseClassifier(object):
...
         pass
 
+    @property
+    @abc.abstractproperty
+    def algorithm_version(self):
+        """Version of the classifier algorithm. The version of the algorithm is
+        matched with the version received as part of job data before training
+        the classifier.
+
+        The version of the algorithm changes everytime the classifier
```
Comment (seanlip) [inline core/classifiers/base.py]: every time

File: core/classifiers/base.py [hunk 26]
```diff
@@ -53,12 +67,13 @@ def type_in_job_result_proto(self):
...
-    def to_dict(self):
-        """Returns a dict representing this classifier.
+    def to_proto(self):
+        """Returns a protobuf of the frozen model consisting of trained
+        parameters.
 
         Returns:
-            dict. A dictionary representation of classifier referred as
-            'classifier_data'. This data is used for prediction.
+            FrozenModel object. A protobuf object of frozen model containing
```
Comment (seanlip) [inline core/classifiers/base.py]: FrozenModel. A protobuf...

File: core/classifiers/base.py [hunk 27]
```diff
@@ -84,11 +99,12 @@ def train(self, training_data):
...
-    def validate(self, classifier_data):
-        """Validates classifier data.
+    def validate(self, frozen_model_proto):
+        """Validates the specified frozen model protobuf object containing
+        parameters of the trained classifier model.
 
         Args:
-            classifier_data: dict of the classifier attributes specific to
-                the classifier algorithm used.
+            frozen_model_proto: FrozenModel object of the frozen model protobuf
```
Comment (seanlip) [inline core/classifiers/base.py]: FrozenModel. The frozen ...

File: core/domain/job_services.py [hunk 28]
```diff
@@ -88,45 +104,35 @@ def train_classifier(algorithm_id, training_data):
...
     """Store result of job in the Oppia server.
 
     Args:
         job_id: str. ID of the job whose result is to be stored.
-        classifier_data: dict. A dictionary representing result of the job.
-
+        algorithm_id: str. ID of the classifier algorithm.
+        frozen_model_proto: Object. A protobuf object that stores trained model
+            parameters.
     Returns:
```
Comment (seanlip) [inline core/domain/job_services.py:132]: Add newline above this.

File: core/domain/job_services.py [hunk 29]
```diff
@@ -88,45 +104,35 @@ def train_classifier(algorithm_id, training_data):
...
-def store_job_result(job_id, classifier_data):
+def store_job_result(job_id, algorithm_id, frozen_model_proto):
     """Store result of job in the Oppia server.
 
     Args:
         job_id: str. ID of the job whose result is to be stored.
-        classifier_data: dict. A dictionary representing result of the job.
-
+        algorithm_id: str. ID of the classifier algorithm.
+        frozen_model_proto: Object. A protobuf object that stores trained model
```
Comment (seanlip) [inline core/domain/job_services.py]: Use a more specific typeinfo, as mentioned elsewhere?

File: core/domain/proto/README.md [hunk 30]
```diff
@@ -0,0 +1,9 @@
+This directory is populated with auto-generated protobuf files by prototool
+upon starting Oppia server.
+It contains various protobuf classes which are used to store trained
```
Comment (seanlip) [inline core/domain/proto/README.md:4]: Add newline above each paragraph

File: core/domain/proto/README.md [hunk 31]
```diff
@@ -0,0 +1,9 @@
...
+classifier model parameters, for example, TextClassifierFrozenModel stores
+paramters of a trained TextClassifier model.
+Since, different classifiers use different protobuf classes, whenever required,
```
Comment (seanlip) [inline core/domain/proto/README.md]: Drop comma after "since"

File: core/domain/proto/README.md [hunk 32]
```diff
@@ -0,0 +1,9 @@
...
+such classes are referred simply as `FrozenModel` throughout the codebase.
+Term `FrozenModel` could be used to refer to any of the protobuf implementation
```
Comment (seanlip) [inline core/domain/proto/README.md]: Term --> The term

Add "s" after implementation

**Review by seanlip — CHANGES_REQUESTED** [review]

File: core/classifiers/CodeClassifier/CodeClassifier.py (hunk 24 above)
Comment (prasanna08) [reply core/classifiers/CodeClassifier/CodeClassifier.py]: Done

File: core/classifiers/base.py (hunk 25 above)
Comment (prasanna08) [reply core/classifiers/base.py]: Done

File: core/classifiers/base.py (hunk 26 above)
Comment (prasanna08) [reply core/classifiers/base.py]: Done

File: core/classifiers/base.py (hunk 27 above)
Comment (prasanna08) [reply core/classifiers/base.py]: Done

File: core/domain/job_services.py (hunk 28 above)
Comment (prasanna08) [reply core/domain/job_services.py:132]: Done

File: core/domain/job_services.py (hunk 29 above)
Comment (prasanna08) [reply core/domain/job_services.py]: Done

File: core/domain/proto/README.md (hunk 30 above)
Comment (prasanna08) [reply core/domain/proto/README.md:4]: Done

File: core/domain/proto/README.md (hunk 31 above)
Comment (prasanna08) [reply core/domain/proto/README.md]: Done

File: core/domain/proto/README.md (hunk 32 above)
Comment (prasanna08) [reply core/domain/proto/README.md]: Done.

File: core/classifiers/CodeClassifier/CodeClassifier.py (hunk 3 above)
Comment (prasanna08) [reply core/classifiers/CodeClassifier/CodeClassifier.py]: Removed the Code Classifier files

File: core/classifiers/CodeClassifier/CodeClassifier.py (hunk 2 above)
Comment (prasanna08) [reply core/classifiers/CodeClassifier/CodeClassifier.py]: Removed the CodeClassifier files.

File: .pylintrc (hunk 1 above)
Comment (prasanna08) [reply .pylintrc:72]: Done

**Comment by prasanna08** [comment]
Hey @seanlip, apologies for delayed response. Addressed the most recent review comments. PTAL

**Comment by seanlip** [comment]
Update: This looks good. We just need to merge the corresponding PR in the oppia/oppia repo first, so placing a hold on this so it doesn't get merged accidentally.

@prasanna08 -- please ping me when you would like this PR to be merged. Thanks!

**Comment by prasanna08** [comment]
Hey @seanlip, can we go ahead and merge this?

**Review by seanlip — APPROVED** [review]
Yup! Sounds good.
